Lukas Mosser
February 2016

In [142]:
import numpy as np

#Excercise 1: Average Permeability fow parallel and perpendicular flow

##1.1: Proof of average permeability for case of parallel flow

We define Darcy's law for a region i with permeability $k_{i}$ and Surface area $S_{i}$ as:
$$q_{i}=\frac{k_{i}}{\mu}S_{i}\frac{dP}{dx}$$
and in the discrete form as:
$$q_{i}=\frac{k_{i}}{\mu}S_{i}\frac{\Delta P}{L}$$

For the case of parallel flow we can assume that the total flow rate $q_{T}$ is equal to the the sum of the individual layer rates $\sum{q_{i}}$

We can therefor write:
$$q_{T}=\sum{q_{i}}$$

Writing Darcy's Law for $q_{T}$:
$$q_{T}= \frac{\overline{k}}{\mu}S_{T}\frac{\Delta P}{L}$$

And therefore subsituting Darcy's Law for the total flow rate and the individual layers:
$$\frac{\overline{k}}{\mu}S_{T}\frac{\Delta P}{L}=\sum{\frac{k_{i}}{\mu}S_{i}\frac{\Delta P}{L}}$$

Cancelling equivalent terms and constants: $L$, $\Delta P$, $\mu$
$$\overline{k}S_{T}=\sum{k_{i}S_{i}}$$
Rearranging terms for the average permeability in case of parallel flow:
$$\overline{k}=\frac{\sum{k_{i}S_{i}}}{S_{T}}$$
Assuming that $S_{T}=\sum{S_{i}}$ we can write the final equation for average permeability in case of parrallel flow:
$$\overline{k}=\frac{\sum{k_{i}S_{i}}}{\sum{S_{i}}}$$

##1.1: Proof of average permeability for case of perpendicular flow

We define Darcy's law for a region i with permeability $k_{i}$ and Surface area $S_{i}$ as:
$$q_{i}=\frac{k_{i}}{\mu}S_{i}\frac{dP}{dx}$$
and in the discrete form as:
$$q_{i}=\frac{k_{i}}{\mu}S_{i}\frac{\Delta P_{i}}{L_{i}}$$

For the case of perpendicular flow we can assume that the total pressure drop $\Delta P_{T}$ is equal to the the sum of the individual pressure drops $\sum{\Delta P_{i}}$
Due to the continuity equation the total flow rate $q_{T}$ is equal to the individual layer flow rates $q_{i}$ and surface area $S_{T}$ is equal to the indivdual layer surface areas $S_{i}$ and is constant for all layers therefore.

Writing Darcy's Law for $q_{T}$:
$$q_{T}=q=\frac{\overline{k}}{\mu}S\frac{\Delta P_{T}}{L_{T}}$$

Assuming constant surface areas and rates we write for the individual layers:
$$q=\frac{k_{i}}{\mu}S\frac{\Delta P_{i}}{L_{i}}$$

Rearranging for pressure drops:
$$\Delta P_{T} =\frac{q \mu L_{T}}{\overline{k}S}$$
$$\Delta P_{i} =\frac{q \mu L_{i}}{k_{i}S}$$

Substituting into for total pressure drops and individual pressure drops:
$${\Delta P_{T}}=\sum{\Delta P_{i}}$$ 
$$\frac{q \mu L_{T}}{\overline{k}S}=\sum{\frac{q \mu L_{i}}{k_{i}S}}$$

Cancelling constant terms: $q$, $S$, $mu$
$$\frac{L_{T}}{\overline{k}}=\sum{\frac{L_{i}}{k_{i}}}$$
And rearranging for the average permeability in case of perpendicular flow:
$$\overline{k}=\frac{L_{T}}{\sum{\frac{L_{i}}{k_{i}}}}$$
This leads us to our final equation for the average permeability in the case of perpendicular flow:
$$\overline{k}=\frac{\sum{L_{i}}}{\sum{\frac{L_{i}}{k_{i}}}}$$

#Excercise 2: Bounds on Average Permeabilities

In [143]:
layer_1_k = np.array([[50, 60, 70, 80, 100], 
                      [55, 65, 45, 35, 150],
                      [60, 68, 100, 120, 200], 
                      [60, 60, 150, 200, 300]])
layer_2_k = np.array([[200, 150, 160, 100, 200], 
                      [220, 100, 90, 80, 280],
                      [250, 200, 120, 90, 400], 
                      [300, 250, 300, 400, 600]])
layer_3_k = np.array([[75, 60, 70, 55, 90], 
                      [80, 50, 40, 35, 130],
                      [90, 80, 65, 65, 180], 
                      [110, 90, 135, 180, 270]])

layer_1_kz = np.array([[5.0,6.0,7.0,8.0,10.0], 
                       [5.5,6.5,4.5,3.5,15.0],
                       [6.0,7.0,10.0,12.0,20.0], 
                       [6.0,6.0,15.0,20.0,30.0]])
layer_2_kz = np.array([[2.0,1.5,1.6,1.0,2.0], 
                       [2.2,1.0,0.9,0.8,2.8],
                       [2.5,2.0,1.2,0.9,4.0], 
                       [3.0,2.5,3.0,4.0,6.0]])
layer_3_kz = np.array([[1.5,1.2,1.4,1.1,1.8], 
                       [1.6,1.0,0.8,0.7,2.6],
                       [1.8,1.6,1.3,1.3,3.6], 
                       [2.2,1.8,2.7,3.6,5.4]])
kxy_layers = np.array([layer_1_k, layer_2_k, layer_3_k])
kz_layers = np.array([layer_1_kz, layer_2_kz, layer_3_kz])

In [144]:
def arithmetic(data):
    data_comp = data.flatten()
    k = 0.
    i = 0.
    for val in data_comp:
        k += val
        i+= 1.
    return k/i

def geometric(data):
    data_comp = data.flatten()
    k = 0.
    i = 0.
    for val in data_comp:
        if val == 0.0:
            return 0.0
        k += np.log(val)
        i+= 1.
    return np.exp(k/i)

def harmonic(data):
    data_comp = data.flatten()
    k = 0.
    i = 0.
    for val in data_comp:
        if val == 0.0:
            return 0.0
        k += 1./val
        i+=1.
    return i/k

def generic_directional_averaging_lower(data, averaging_method=arithmetic, dir="x"):
    averages = []
    if dir=="x":
        for i in range(data.shape[0]):
            for j in range(data.shape[1]):
                averages.append(averaging_method(data[i, j, :]))
    elif dir=="y":
        for i in range(data.shape[0]):
            for j in range(data.shape[2]):
                averages.append(averaging_method(data[i, :, j]))
    elif dir=="z":
        for i in range(data.shape[1]):
            for j in range(data.shape[2]):
                averages.append(averaging_method(data[:, i, j]))
    return np.array(averages)

def generic_directional_averaging_upper(data, averaging_method=arithmetic, dir="x"):
    averages = []
    if dir=="x":
        for i in range(data.shape[2]):
                averages.append(averaging_method(data[:, :, i]))
    elif dir=="y":
        for i in range(data.shape[1]):
                averages.append(averaging_method(data[:, i, :]))
    elif dir=="z":
        for i in range(data.shape[0]):
                averages.append(averaging_method(data[i, :, :]))
    return np.array(averages)

def cardwell_parsons(data, dir="x", bound="lower"):
    if bound == "lower":
        return arithmetic(generic_directional_averaging_lower(data, averaging_method=harmonic, dir=dir))
    elif bound == "upper":
        return harmonic(generic_directional_averaging_upper(data, averaging_method=arithmetic, dir=dir))

In [145]:
k_arith_xy = arithmetic(kxy_layers)
k_arith_z = arithmetic(kz_layers)
print k_arith_xy, k_arith_z

k_geom_xy = geometric(kxy_layers)
k_geom_z = geometric(kz_layers)
print k_geom_xy, k_geom_z

k_harm_xy = harmonic(kxy_layers)
k_harm_z = harmonic(kz_layers)
print k_harm_xy, k_harm_z

141.133333333 4.78166666667
112.010351441 3.04189797579
92.012156271 2.14895296487


##2.1: Arithmetic Harmonic and Geometrc Averages of the data:
For the horizontal permeability $k_{h}$ the arithmetic, geometric and harmonic averages are:
$$k_{h, harm} = 92.0 \ [mD]$$
$$k_{h, geom} = 112.0 \ [mD]$$
$$k_{h, arith} = 141.1 \ [mD]$$
For the vertical permeability $k_{z}$ the arithmetic, geometric and harmonic averages are:
$$k_{z, harm} = 2.15 \ [mD]$$
$$k_{z, geom} = 3.04 \ [mD]$$
$$k_{z, arith} = 4.78 \ [mD]$$

In [146]:
from itertools import product

for comb in product(["upper", "lower"], ["x", "y"]):
    bound = comb[0]
    dir = comb[1]
    print "Direction "+dir+" Bound: "+bound+" k average "+str(cardwell_parsons(kxy_layers, dir=dir, bound=bound))

Direction x Bound: upper k average 128.642204826
Direction y Bound: upper k average 125.917252914
Direction x Bound: lower k average 119.045514943
Direction y Bound: lower k average 120.164955901


In [147]:
for comb in product(["upper", "lower"], ["z"]):
    bound = comb[0]
    dir = comb[1]
    print "Direction "+dir+" Bound: "+bound+" k average "+str(cardwell_parsons(kz_layers, dir=dir, bound=bound))

Direction z Bound: upper k average 2.83882035692
Direction z Bound: lower k average 2.79792093858


##2.2: Wiener and Cardwell Parsons Bounds for $K_{x}$, $K_{y}$, $K_{z}$
The Wiener Bounds are given as:
$$K_{i, harm}<K_{i}<K_{i, arith} \ i = h, \ z$$
Therefore the Wiener Bounds for horizontal and vertical directions are:
$$92.0<K_{h}<141.1 \ [mD]$$
$$2.15<K_{z}<4.78 \ [mD]$$

The Cardwell Parson Bounds are given as:
$$[A, \ H](K_{i})<K_{i}<[H,\ A](K_{i}), \ i = x, \ y, \ z$$
Therefore the Cardwell Parsons Bounds for x,y, and z directions are:
$$119.0<K_{x}<128.6 [mD]$$
$$120.2<K_{y}<125.9 [mD]$$
$$2.79<K_{z}<2.84 [mD]$$

##2.3: Cardwell Parsons Estimates for $K_{x}$, $K_{y}$, $K_{z}$

In [148]:
K_x_cp = np.sqrt(cardwell_parsons(kxy_layers, dir="x", bound="lower")*cardwell_parsons(kxy_layers, dir="x", bound="upper"))
K_y_cp = np.sqrt(cardwell_parsons(kxy_layers, dir="y", bound="lower")*cardwell_parsons(kxy_layers, dir="y", bound="upper"))
K_z_cp = np.sqrt(cardwell_parsons(kz_layers, dir="z", bound="lower")*cardwell_parsons(kz_layers, dir="z", bound="upper"))
print K_x_cp, K_y_cp, K_z_cp

123.750868752 123.007484096 2.81829645664


The Cardwell Parsons Estimate is given as:
$$K_{i, cp}=\sqrt{[A, \ H](K_{i})[H,\ A](K_{i})}, \ i = x, \ y, \ z$$
Therefore the Cardwell Parsons Estimates for $K_{x}$, $K_{y}$, $K_{z}$ are:
$$K_{x, cp} = 123.75 \ [mD]$$
$$K_{y, cp} =123.01 \ [mD]$$
$$K_{z, cp} = 2.82 \ [mD]$$

##2.4: Questions 2.1, 2.2 and 2.3 when replacing all values in layer 2 for $K_{z}$ with 0.

In [149]:
kz_layers[1, :, :] =np.array([[0.0,0.0,0.0,0.0,0.0], 
                              [0.0,0.0,0.0,0.0,0.0],
                              [0.0,0.0,0.0,0.0,0.0], 
                              [0.0,0.0,0.0,0.0,0.0]])
k_arith_xy = arithmetic(kxy_layers)
k_arith_z = arithmetic(kz_layers)
print k_arith_xy, k_arith_z

k_geom_xy = geometric(kxy_layers)
k_geom_z = geometric(kz_layers)
print k_geom_xy, k_geom_z

k_harm_xy = harmonic(kxy_layers)
k_harm_z = harmonic(kz_layers)
print k_harm_xy, k_harm_z

141.133333333 4.03333333333
112.010351441 0.0
92.012156271 0.0


###2.4.1: Arithmetic Harmonic and Geometrc Averages of the data, zero layer 2 in $K_{z}$:
For the horizontal permeability $k_{h}$ the arithmetic, geometric and harmonic averages are:
$$k_{h, harm} = 92.0 \ [mD]$$
$$k_{h, geom} = 112.0 \ [mD]$$
$$k_{h, arith} = 141.1 \ [mD]$$
For the vertical permeability $k_{z}$ the arithmetic, geometric and harmonic averages are:
$$k_{z, harm} = 0.0 \ [mD]$$
$$k_{z, geom} = 0.0 \ [mD]$$
$$k_{z, arith} = 4.03 \ [mD]$$

In [150]:
for comb in product(["upper", "lower"], ["x", "y"]):
    bound = comb[0]
    dir = comb[1]
    print "Direction "+dir+" Bound: "+bound+" k average "+str(cardwell_parsons(kxy_layers, dir=dir, bound=bound))
for comb in product(["upper", "lower"], ["z"]):
    bound = comb[0]
    dir = comb[1]
    print "Direction "+dir+" Bound: "+bound+" k average "+str(cardwell_parsons(kz_layers, dir=dir, bound=bound))

Direction x Bound: upper k average 128.642204826
Direction y Bound: upper k average 125.917252914
Direction x Bound: lower k average 119.045514943
Direction y Bound: lower k average 120.164955901
Direction z Bound: upper k average 0.0
Direction z Bound: lower k average 0.0


####2.4.2 Wiener and Cardwell Parsons Bounds for $K_{x}$, $K_{y}$, $K_{z}$
The Wiener Bounds are given as:
$$K_{i, harm}<K_{i}<K_{i, arith} \ i = h, \ z$$
Therefore the Wiener Bounds for horizontal and vertical directions are:
$$92.0<K_{h}<141.1 \ [mD]$$
$$0.0<K_{z}<4.03 \ [mD]$$
The Cardwell Parson Bounds are given as:
$$[A, \ H](K_{i})<K_{i}<[H,\ A](K_{i}), \ i = x, \ y, \ z$$
Therefore the Cardwell Parsons Bounds for x,y, and z directions are:
$$119.0<K_{x}<128.6 [mD]$$
$$120.2<K_{y}<125.9 [mD]$$
$$0.0<K_{z}<0.0 [mD]$$

###2.4.3 Cardwell Parsons Estimates for $K_{x}$, $K_{y}$, $K_{z}$

In [151]:
K_x_cp = np.sqrt(cardwell_parsons(kxy_layers, dir="x", bound="lower")*cardwell_parsons(kxy_layers, dir="x", bound="upper"))
K_y_cp = np.sqrt(cardwell_parsons(kxy_layers, dir="y", bound="lower")*cardwell_parsons(kxy_layers, dir="y", bound="upper"))
K_z_cp = np.sqrt(cardwell_parsons(kz_layers, dir="z", bound="lower")*cardwell_parsons(kz_layers, dir="z", bound="upper"))
print K_x_cp, K_y_cp, K_z_cp

123.750868752 123.007484096 0.0


The Cardwell Parsons Estimate is given as:
$$K_{i, cp}=\sqrt{[A, \ H](K_{i})[H,\ A](K_{i})}, \ i = x, \ y, \ z$$
Therefore the Cardwell Parsons Estimates for $K_{x}$, $K_{y}$, $K_{z}$ are:
$$K_{x, cp} = 123.75 \ [mD]$$
$$K_{y, cp} =123.01 \ [mD]$$
$$K_{z, cp} = 0.0 \ [mD]$$

##2.5 Check the inequalities
The inequalities for the case of zero permeability in layer two will result in the lower Cardwell Parson bound to become zero.
The Wiener bounds and Cardwell Parson bounds are still fullfilled in all cases as the lower wiener bound is zero as well as the lower cardwell parson bound. The cardwell parson upper bound is smaller than the upper wiener bound.

#Excercise 3: Half Cell Transmissivities

In [152]:
k_coarse_green = np.array([50, 100, 200, 50])
k_coarse_red = np.array([100, 50, 500, 0])
k_coarse_blue = np.array([80, 100, 50, 800])

k_coarse_green = np.array([50, 100, 200, 50])
k_coarse_red = np.array([100, 50, 500, 0])
k_coarse_blue = np.array([80, 100, 50, 800])

k_coarse_green_1 = np.array([50, 100])
k_coarse_green_2 = np.array([200, 50])

k_coarse_red_1 = np.array([100, 50])
k_coarse_red_2 = np.array([500, 0])

k_coarse_blue_1 = np.array([80, 100])
k_coarse_blue_2 = np.array([50, 800])

##3.1: Compute the mean coarse permeabilities

In [153]:
print harmonic(k_coarse_green), harmonic(k_coarse_red), harmonic(k_coarse_blue)

72.7272727273 0.0 91.4285714286


The mean coarse permeabilities for $K_{green}$, $K_{red}$, $K_{blue}$ are:
$$K_{green} = 72.7 \ [mD]$$
$$K_{red} =0.0 \ [mD]$$
$$K_{blue} =91.4 \ [mD]$$

##3.1: Compute the mean half coarse cell permeabilities

In [154]:
print harmonic(k_coarse_green_1), harmonic(k_coarse_green_2)
print harmonic(k_coarse_red_1), harmonic(k_coarse_red_2)
print harmonic(k_coarse_blue_1), harmonic(k_coarse_blue_2)

66.6666666667 80.0
66.6666666667 0.0
88.8888888889 94.1176470588


The mean coarse permeabilities for 
$K_{green, 1}$, $K_{green, 2}$,  $K_{red, 1}$, $K_{red, 2}$,  $K_{blue, 1}$, $K_{blue, 2}$are:
$$K_{green, 1} = 66.7 \ [mD]$$
$$K_{green, 2} = 80.0 \ [mD]$$
$$K_{red, 1} =66.7 \ [mD]$$
$$K_{red, 2} =0.0 \ [mD]$$
$$K_{blue, 1} =88.9 \ [mD]$$
$$K_{blue, 2} =94.1 \ [mD]$$

##3.3: Transmissibilities green<->red<->blue for whole coarse and half coarse cells

We assume that surface areas between cells and distances between center nodes are the same for all cells, therefore the computation of the transmissivity reduces to the harmonic average between the corse cells and the half coarse cells adjacent to cell boundaries.

In [155]:
T_green_red_coarse_full = harmonic(np.array([k_coarse_green, k_coarse_red]))
T_red_blue_coarse_full = harmonic(np.array([k_coarse_red, k_coarse_blue]))
print T_green_red_coarse_full, T_red_blue_coarse_full

0.0 0.0


In [156]:
T_green_red_coarse_half = harmonic(np.array([k_coarse_green_2, k_coarse_red_1]))
T_red_blue_coarse_half = harmonic(np.array([k_coarse_red_2, k_coarse_blue_1]))
print T_green_red_coarse_half, T_red_blue_coarse_half

72.7272727273 0.0


The transmissivities for the full coarse cells is:
$$T_{green, red, full} = 0.0 \ [mD ft]$$
$$T_{red, blue, full} = 0.0 \ [mD ft]$$

The transmissivities for the half coarse cells is:
$$T_{green, red, half} = 72.7 \ [mD ft]$$
$$T_{red, blue, half} = 0.0 \ [mD ft]$$

##3.4: Conclusion of excercise 3

There is a discrepancy between the half coarse and full coarse cell transmissivities.
Using the full coarse cell, due to the harmonic averaging all transmissivities of cells bordering a zero valued permeability cell will have zero transmissivity. Whereas using half coarse cells we observe that the transmissivity between green and red is not equal to zero, while red and blue remains zero. Therefore we conclude that whenever the permeability of one cell is equal to zero this will always lead to a zero transmissivity between cells. Only if both half cells, or full cells have permeabilities not equal zero, will there be a non zero transmissivities between the cells.